File: 18isUnivalueList.js

// iterative
// n = number of nodes
// time:  O(n)
// space: O(1)

const isUnivalueList = (head) => {
  // todo
  let val = head.val;
  let current = head;
  
  while (current) {
    if (current.val !== val) return false;
    current = current.next;
  }
  
  return true;
};

// ------------------------- //
// my implementation
// recursive
// n = number of nodes
// time:  O(n)
// space: O(n)

const isUnivalueList = (head, prevVal = head.val) => {
  // todo
  if (!head) return true;
  if (head.val !== prevVal) return false;
  
  return isUnivalueList(head.next, head.val);
};



// -------------------------------------- //
// structy
// recursive
// n = number of nodes
// time:  O(n)
// space: O(n)


const isUnivalueList = (head, prevVal = null) => {
  if (head === null) return true;
  if (prevVal === null || prevVal === head.val) {
    return isUnivalueList(head.next, head.val);
  } else {
    return false;
  }
}

File: Python/18is_univalue_list.py


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)

def is_univalue_list(head):
  current = head
  while current is not None:
    if current.val != head.val:
      return False
    current = current.next
  return True


# -----------------------------------------
# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)

def is_univalue_list(head, prev_val=None):
  if head is None:
    return True
  if prev_val is None or head.val == prev_val:
    return is_univalue_list(head.next, head.val)
  else:
    return False

# -----------------------------------------
# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)

def is_univalue_list(head, value= None):
  if head is None:
    return True
  if value is not None and head.val != value:
    return False

  return is_univalue_list(head.next, head.val)
